### PR TITLE
Package `yarp` with setuptools

### PR DESCRIPTION
**Pull request by diegoferigo**
This PR follows the effort to bring some of our projects to the [Python Package Index](https://pypi.org/) (see [my packages](https://pypi.org/user/dgferigo/) to get an overview).

More in details:

- Add a `setup.py` and `pyproject.toml` files to build both source distributions and wheels with [PEP517](https://www.python.org/dev/peps/pep-0517/) and [PEP518](https://www.python.org/dev/peps/pep-0518/) support ([explanation](https://grassfedcode.medium.com/pep-517-and-518-in-plain-english-47208ca8b7a6)).
- Add a new python workflow to generate the Python packages and upload them to PyPI.
- Minor modifications to the CMake logic.

Only GNU/Linux wheels are built with this PR. The wheels comply with the most recent [PEP600](https://www.python.org/dev/peps/pep-0600/), particularly `manylinux_2_24`. The wheels are self-contained, and they do not need any runtime dependency (similarly to https://github.com/robotology/idyntree/pull/844).

The process exploits [`diegoferigo/cmake-build-extension`](https://github.com/diegoferigo/cmake-build-extension) that acts as a bridge between CMake projects and setuptools.

Questions and comments:

- I created the package so that its usage is 100% compatible with the existing bindings:
  ```python
  import yarp
  b = yarp.Bottle()
  ```
  However, the Python package has a completely different structure since it is a proper package that lives in a `yarp/` subfolder with an `__init__.py` file. It also ships the entire CMake project so that it can be consumed by downstream packages. Their wheels, in fact, can import the targets and link against yarp automatically while building other PEP600 compliant packages.
- ~Do you want to maintain this pipeline? If not, we can maybe disable the automatic PyPI upload and I can do it manually with `twine` when needed.~ Agreed with @drdanz to add the pip / PyPI resources in a subfolder with @diegoferigo as CODEOWNER.
- ~The `yarp` name is [already taken](https://pypi.org/project/yarp/) :disappointed:  It does not seem really updated nor maintained, I contacted the owned but he's still interested in keeping the name. Therefore, we have to select a new name, maybe `yarp-middleware`?~ Used `yarp_middleware`.
- ~Any better way to quickly install `ycm` in debian stretch (necessary to build PEP600 wheels)? Right now I use the new official ycm-cmake-modules packages but in stretch it results in missing dependencies. However, the package is installed and its files correctly found.~ See https://github.com/robotology/ycm/pull/380.
- `yarp` cannot be build in debian stretch with the default gcc due to [this](http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#1684), I had to rely on clang. Though, also old clang versions fails to build yarp (see details below), I had to use clang7.
- ~The package would include also the yarp binaries but they will not get exposed automatically. The users has to add the `/path/to/site-packages/yarp/bin` to their path. This is a limitation that I don't know how to solve, and I'm not sure if it is supported by the official tooling. For Python scripts, there's the [`entry_point`](https://setuptools.readthedocs.io/en/latest/userguide/entry_point.html) option that exposes them to the user, but it does not work for C++ executables. A workaround would be creating Python wrapper that call the C++ executables, but maybe we can do it in a future PR if we want to get this additional feature.~ New binaries support introduced in https://github.com/diegoferigo/cmake-build-extension/pull/14.
- As documented [here](https://pybind11.readthedocs.io/en/stable/compiling.html#pep-518-requirements-pip-10-required), a recent version of pip is necessary for both PEP518 (isolated builds) and PEP600 (perennial manylinux wheels) support.

<details>
<summary>Clang error</summary>

```
 /usr/bin/clang++-4.0 -DBUILDING_YARP -DYARP_HAS_ACE -D__ACE_INLINE__ -I../../src/libYARP_companion/src -I../../src/libYARP_os/src -I../../src/libYARP_conf/src -Isrc/libYARP_conf/src -Wall -Wextra -Wsign-compare -Wpointer-arith -Winit-self -Wnon-virtual-dtor -Wcast-align -Wunused -Wvla -Wmissing-include-dirs -Wreorder -Wsizeof-pointer-memaccess -Woverloaded-virtual -Wtautological-undefined-compare -Wmismatched-new-delete -Wparentheses-equality -Wundef -Wredundant-decls -Wunknown-pragmas -Wunused-result -Wheader-guard -Wignored-attributes -Wnewline-eof -Wdangling-else -Wgcc-compat -Wmicrosoft-exists -Wstatic-inline-explicit-instantiation -Wtautological-compare -Winconsistent-missing-override -Wnull-conversion  -Wno-unused-parameter  -Wdeprecated-declarations  -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -std=c++14 -MD -MT src/libYARP_companion/src/CMakeFiles/YARP_companion.dir/yarp/companion/impl/Companion.cmdPray.cpp.o -MF src/libYARP_companion/src/CMakeFiles/YARP_companion.dir/yarp/companion/impl/Companion.cmdPray.cpp.o.d -o src/libYARP_companion/src/CMakeFiles/YARP_companion.dir/yarp/companion/impl/Companion.cmdPray.cpp.o -c ../../src/libYARP_companion/src/yarp/companion/impl/Companion.cmdPray.cpp
    ../../src/libYARP_companion/src/yarp/companion/impl/Companion.cmdPray.cpp:83:9: error: 'va_start' used in function with fixed args
            va_start(args, format);
            ^
    /usr/include/clang/4.0.1/include/stdarg.h:33:29: note: expanded from macro 'va_start'
    #define va_start(ap, param) __builtin_va_start(ap, param)
                                ^
    1 error generated.
```

</details>

TODO:

- [x] Wait new release of cibuildwheel. Now we install it from sources to get the pypa/build support. (we can use the pre-release `>= v2.0.0a4`)
- [x] Wait https://github.com/robotology/ycm/pull/380 and update pyproject.toml

### CODE REVIEW

**Comment by diegoferigo** [comment]
cc @drdanz @traversaro This is now ready for review, modulo commit rebasing. It does not strictly depend on https://github.com/robotology/ycm/pull/380 since this PR uses the ycm package built from the repository and not downloaded from PyPI. We can even merge https://github.com/robotology/ycm/pull/380 without the PyPI deployment while waiting the finalization of the name request (https://github.com/pypa/pypi-support/issues/1096) that could take a while.

**Comment by diegoferigo** [comment]
In the last build, CMake's install process loops forever in the bindings folder. I managed to debug the problem and it seems related to the following lines:

https://github.com/robotology/yarp/blob/5980201205e143b34e9a2c4485b1d255dfeb0493/bindings/CMakeLists.txt#L128-L130

In short, CMake wants to install all the swig folders (`bindings/*`), but when packaging for Python with the resources of this PR (stored in `bindings/python/pip`) the build folder used by the process is stored into `bindings/python/pip/build`. This location that cannot be changed causes the infinite loop occurring in CI.

@drdanz @traversaro do you have any suggestion about how to overcome it? Is it necessary installing the SWIG resources of the bindings? This is the last blocker for this PR.

**Comment by traversaro** [comment]
Can't we just move `bindings/python/pip` to `bindings/python-pip` or `scripts/pip` ?

**Comment by traversaro** [comment]
> 
> Can't we just move `bindings/python/pip` to `bindings/python-pip` or `scripts/pip` ?

The alternative is modified https://github.com/robotology/yarp/blob/5980201205e143b34e9a2c4485b1d255dfeb0493/bindings/CMakeLists.txt#L128-L130 to manually exclude the `pip` directory, but just changing directory seems easier.

**Comment by diegoferigo** [comment]
We could actually use `EXCLUDE PATTERN` of CMake's [install](https://cmake.org/cmake/help/latest/command/install.html#directory), but I agree that changing directory is simpler. I moved the resources to `scripts/pip` in https://github.com/robotology/yarp/compare/468eb594ed2a75d8caaf8aa097bd3957a2259224..e2ae22e7d68c48ebdc158bed345526deab08104d.

**Comment by diegoferigo** [comment]
There's no need to link against the Python libs. It [causes](https://github.com/robotology/yarp/pull/2573/checks?check_run_id=2929907009#step:6:4764) the following problems:

```
       + sh -c 'python -c '"'"'import yarp; yarp.Bottle()'"'"''
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/tmp/tmp.iAb7pwgXGd/venv/lib/python3.7/site-packages/yarp/__init__.py", line 1, in <module>
      from .bindings.yarp import *
    File "/tmp/tmp.iAb7pwgXGd/venv/lib/python3.7/site-packages/yarp/bindings/yarp.py", line 13, in <module>
      from . import _yarp
  ImportError: /tmp/tmp.iAb7pwgXGd/venv/lib/python3.7/site-packages/yarp/bindings/_yarp.so: undefined symbol: PyCMethod_New
                                                              ✕ 8.80s
Error: Command ['sh', '-c', "python -c 'import yarp; yarp.Bottle()'"] failed with code 1. 
```

Fixed in 064c562.

**Comment by diegoferigo** [comment]
There's no need to link against the Python libs. It [causes](https://github.com/robotology/yarp/pull/2573/checks?check_run_id=2929907009#step:6:4764) the following problems:

```
       + sh -c 'python -c '"'"'import yarp; yarp.Bottle()'"'"''
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/tmp/tmp.iAb7pwgXGd/venv/lib/python3.7/site-packages/yarp/__init__.py", line 1, in <module>
      from .bindings.yarp import *
    File "/tmp/tmp.iAb7pwgXGd/venv/lib/python3.7/site-packages/yarp/bindings/yarp.py", line 13, in <module>
      from . import _yarp
  ImportError: /tmp/tmp.iAb7pwgXGd/venv/lib/python3.7/site-packages/yarp/bindings/_yarp.so: undefined symbol: PyCMethod_New
                                                              ✕ 8.80s
Error: Command ['sh', '-c', "python -c 'import yarp; yarp.Bottle()'"] failed with code 1. 
```

`Fixed in 064c562.~

Edit: In theory there's not need, but in the yarp case this is required due to the following:

https://github.com/robotology/yarp/blob/5980201205e143b34e9a2c4485b1d255dfeb0493/bindings/yarp.i#L268

**Comment by diegoferigo** [comment]
I just realized that cibuildwheel [finds the wrong Python version](https://github.com/robotology/yarp/pull/2573/checks?check_run_id=2930314839#step:6:1117):

```
-- Found Python3: /usr/local/bin/python3.10 (found version "3.10.0") found components: Interpreter Development.Module 
```

Passing explicitly the active interpreter location as we do in other projects should solve it.

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -82,3 +82,4 @@
 /src/yarpviz/**                                  @Nicogene
 CMakeLists.txt                                   @drdanz
 *.cmake                                          @drdanz
+scripts/pip                                      @dferigo
```
Comment (drdanz) [inline .github/CODEOWNERS:85]: I might be wrong here, but I think that you might need to do this:
```suggestion
scripts/pip/**                                   @dferigo
```

Comment (diegoferigo) [reply .github/CODEOWNERS]: Right, done in https://github.com/robotology/yarp/compare/cb1f143b3deeabdd0c9cdafacc37f56222d8bfa8..27d14b3300ed3bee0824c5b0a8e0b4c6608fa7bd

**Comment by diegoferigo** [comment]
@drdanz This PR is finalized now. I rebased on top of `master`, let's wait CI to turn green and then from my side it can be merged. 

Similarly to YCM, also here there's the same process for the PyPI token. I manually [uploaded](https://pypi.org/project/yarp-middleware/) with `twine` a locally built version and added you as owner. 

I think that the Sdist hack to support subfolders is related to the missing README in PyPI, also YCM is affected. I'll check how to fix it, maybe on the [`cmake-build-extension`](https://github.com/diegoferigo/cmake-build-extension) level.

**Review by drdanz — CHANGES_REQUESTED** [review]
Thanks @diegoferigo! LGTM, but there is just a small issue. Please fix this ASAP, and I'll merge this PR before YARP 3.5 release

File: doc/release/v3_4_5.md [hunk 2]
```diff
@@ -32,6 +32,13 @@ Bug Fixes
   to `IBattery` interface are not implemented (returning false).
   The only mandatory  method is `getBatteryStatus()`.
 
+New Features
+------------
+
+### Bindings
+
+* Python: introduced unofficial support to install YARP with tools like `pip`. A new `yarp-middleware` package is now distributed in PyPI.
+
```
[the remark this answers is not in the document]
Comment (diegoferigo) [reply doc/release/v3_4_5.md]: I have no idea how this new process works (did I use the right sections?), but I tried to adapt what I already found in that folder. Is there any documentation about it?

[the remark this answers is not in the document]
Comment (drdanz) [reply doc/release/v3_4_5.md]: It's perfect now. Thanks!

**Review by drdanz — APPROVED** [review]

**Comment by drdanz** [comment]
Merged, thanks.

**Comment by diegoferigo** [comment]
With the fix of https://github.com/robotology/yarp/pull/2636, the continuous delivery pipeline to PyPI seems working fine:

https://pypi.org/project/yarp-middleware/

**Pull request merged**